File: backend/app/crud/transaction.py
from typing import List, Optional, Dict, Any
from datetime import date, datetime
from sqlalchemy.orm import Session
from sqlalchemy import and_, or_, func, desc, asc

from app.models.transaction import Transaction
from app.schemas.transaction import TransactionCreate, TransactionUpdate

def get_transaction(db: Session, transaction_id: int, tenant_id: int) -> Optional[Transaction]:
    """Get a single transaction by ID"""
    return db.query(Transaction).filter(
        and_(
            Transaction.id == transaction_id,
            Transaction.tenant_id == tenant_id
        )
    ).first()

def get_transaction_by_user(
    db: Session,
    transaction_id: int,
    user_id: int,
    tenant_id: int
) -> Optional[Transaction]:
    """Get a transaction by ID for a specific user"""
    return db.query(Transaction).filter(
        and_(
            Transaction.id == transaction_id,
            Transaction.user_id == user_id,
            Transaction.tenant_id == tenant_id
        )
    ).first()

def get_transactions_by_user(
    db: Session,
    *,
    user_id: int,
    tenant_id: int,
    skip: int = 0,
    limit: int = 100,
    filters: Optional[Dict[str, Any]] = None
) -> List[Transaction]:
    """Get transactions for a user with optional filters"""
    query = db.query(Transaction).filter(
        and_(
            Transaction.user_id == user_id,
            Transaction.tenant_id == tenant_id
        )
    )
    
    if filters:
        if filters.get('type'):
            query = query.filter(Transaction.type == filters['type'])
        if filters.get('category_id'):
            query = query.filter(Transaction.category_id == filters['category_id'])
        if filters.get('account_id'):
            query = query.filter(Transaction.account_id == filters['account_id'])
        if filters.get('date_from'):
            query = query.filter(Transaction.transaction_date >= filters['date_from'])
        if filters.get('date_to'):
            query = query.filter(Transaction.transaction_date <= filters['date_to'])
        if filters.get('amount_min'):
            query = query.filter(Transaction.amount >= filters['amount_min'])
        if filters.get('amount_max'):
            query = query.filter(Transaction.amount <= filters['amount_max'])
        if filters.get('search'):
            search_term = f"%{filters['search']}%"
            query = query.filter(
                or_(
                    Transaction.description.ilike(search_term),
                    Transaction.reference_number.ilike(search_term)
                )
            )
    
    return query.order_by(desc(Transaction.transaction_date)).offset(skip).limit(limit).all()

def get_monthly_summary(
    db: Session,
    *,
    user_id: int,
    tenant_id: int,
    year: int,
    month: int
) -> Dict[str, Any]:
    """Get monthly transaction summary for a user"""
    start_date = date(year, month, 1)
    if month == 12:
        end_date = date(year + 1, 1, 1)
    else:
        end_date = date(year, month + 1, 1)
    
    query = db.query(Transaction).filter(
        and_(
            Transaction.user_id == user_id,
            Transaction.tenant_id == tenant_id,
            Transaction.transaction_date >= start_date,
            Transaction.transaction_date < end_date
        )
    )
    
    total_credits = query.filter(Transaction.type == 'credit').with_entities(
        func.sum(Transaction.amount)
    ).scalar() or 0
    
    total_debits = query.filter(Transaction.type == 'debit').with_entities(
        func.sum(Transaction.amount)
    ).scalar() or 0
    
    return {
        'total_credits': total_credits,
        'total_debits': total_debits,
        'net_amount': total_credits - total_debits,
        'transaction_count': query.count()
    }

def get_category_breakdown(
    db: Session,
    *,
    user_id: int,
    tenant_id: int,
    date_from: Optional[date] = None,
    date_to: Optional[date] = None
) -> List[Dict[str, Any]]:
    """Get transaction breakdown by category"""
    query = db.query(
        Transaction.category_id,
        func.sum(Transaction.amount).label('total_amount'),
        func.count(Transaction.id).label('transaction_count')
    ).filter(
        and_(
            Transaction.user_id == user_id,
            Transaction.tenant_id == tenant_id
        )
    )
    
    if date_from:
        query = query.filter(Transaction.transaction_date >= date_from)
    if date_to:
        query = query.filter(Transaction.transaction_date <= date_to)
    
    return query.group_by(Transaction.category_id).all()

def get_account_summary(
    db: Session,
    *,
    account_id: int,
    user_id: int
) -> Dict[str, Any]:
    """Get transaction summary for a specific account"""
    query = db.query(Transaction).filter(
        and_(
            Transaction.account_id == account_id,
            Transaction.user_id == user_id
        )
    )
    
    total_credits = query.filter(Transaction.type == 'credit').with_entities(
        func.sum(Transaction.amount)
    ).scalar() or 0
    
    total_debits = query.filter(Transaction.type == 'debit').with_entities(
        func.sum(Transaction.amount)
    ).scalar() or 0
    
    transaction_count = query.count()
    
    last_transaction = query.order_by(desc(Transaction.transaction_date)).first()
    last_transaction_date = last_transaction.transaction_date if last_transaction else None
    
    return {
        'transaction_count': transaction_count,
        'total_credits': total_credits,
        'total_debits': total_debits,
        'last_transaction_date': last_transaction_date
    }

def count_by_account(db: Session, account_id: int) -> int:
    """Count transactions for a specific account"""
    return db.query(Transaction).filter(Transaction.account_id == account_id).count()

def create_transaction(
    db: Session,
    *,
    obj_in: TransactionCreate,
    user_id: int,
    tenant_id: int
) -> Transaction:
    """Create a new transaction"""
    db_obj = Transaction(
        **obj_in.dict(),
        user_id=user_id,
        tenant_id=tenant_id
    )
    db.add(db_obj)
    db.commit()
    db.refresh(db_obj)
    return db_obj

def create(
    db: Session,
    *,
    obj_in: TransactionCreate
) -> Transaction:
    """Create a new transaction (alternative method)"""
    db_obj = Transaction(**obj_in.dict())
    db.add(db_obj)
    db.commit() 
    db.refresh(db_obj)
    return db_obj

def get_by_recurring_and_date(
    db: Session,
    *,
    recurring_id: int,
    transaction_date: date
) -> Optional[Transaction]:
    """Check if a transaction already exists for this recurring template and date"""
    return db.query(Transaction).filter(
        and_(
            Transaction.recurring_template_id == recurring_id,
            Transaction.transaction_date == transaction_date
        )
    ).first()

def update_transaction(
    db: Session,
    *,
    db_obj: Transaction,
    obj_in: TransactionUpdate
) -> Transaction:
    """Update an existing transaction"""
    update_data = obj_in.dict(exclude_unset=True)
    for field, value in update_data.items():
        setattr(db_obj, field, value)
    
    db.add(db_obj)
    db.commit()
    db.refresh(db_obj)
    return db_obj

def delete_transaction(
    db: Session,
    *,
    transaction_id: int,
    user_id: int,
    tenant_id: int
) -> Optional[Transaction]:
    """Delete a transaction"""
    db_obj = db.query(Transaction).filter(
        and_(
            Transaction.id == transaction_id,
            Transaction.user_id == user_id,
            Transaction.tenant_id == tenant_id
        )
    ).first()
    
    if db_obj:
        db.delete(db_obj)
        db.commit()
        return db_obj
    return None

File: frontend/src/types/Transaction.ts
// Transaction types and enums matching backend schema
export type TransactionType = 'credit' | 'debit';
export type TransactionCategoryType = 'income' | 'expense' | 'transfer';
export type RecurringFrequency = 'daily' | 'weekly' | 'monthly' | 'quarterly' | 'yearly';

// Core Transaction interface
export interface Transaction {
  id: number;
  tenant_id: number;
  user_id: number;
  type: TransactionType;
  amount: number; // Decimal from backend as number in frontend
  currency: string;
  description?: string;
  transaction_date: string; // ISO date string
  category_id?: number;
  subcategory_id?: number;
  account_id?: number;
  reference_number?: string;
  is_recurring: boolean;
  recurring_template_id?: number;
  attachments?: Record<string, unknown>;
  extra_data?: Record<string, unknown>;
  created_at: string; // ISO string
  updated_at: string; // ISO string
}

// Transaction creation interface
export interface TransactionCreate {
  type: TransactionType;
  amount: number;
  currency?: string;
  description?: string;
  transaction_date: string; // ISO date string
  category_id?: number;
  subcategory_id?: number;
  account_id?: number;
  reference_number?: string;
  is_recurring?: boolean;
  recurring_template_id?: number;
  attachments?: Record<string, unknown>;
  extra_data?: Record<string, unknown>;
}

// Transaction update interface
export interface TransactionUpdate {
  type?: TransactionType;
  amount?: number;
  currency?: string;
  description?: string;
  transaction_date?: string; // ISO date string
  category_id?: number;
  subcategory_id?: number;
  account_id?: number;
  reference_number?: string;
  is_recurring?: boolean;
  recurring_template_id?: number;
  attachments?: Record<string, unknown>;
  extra_data?: Record<string, unknown>;
}

// Transaction with details (includes related entities)
export interface TransactionWithDetails extends Transaction {
  category?: TransactionCategory;
  subcategory?: TransactionSubcategory;
  account?: FinancialAccount;
  recurring_template?: RecurringTransaction;
}

// Transaction Category interface
export interface TransactionCategory {
  id: number;
  tenant_id: number;
  user_id?: number;
  name: string;
  type?: TransactionCategoryType;
  color?: string; // Hex color code
  icon?: string;
  is_system: boolean;
  created_at: string; // ISO string
}

// Transaction Category creation interface
export interface TransactionCategoryCreate {
  name: string;
  type?: TransactionCategoryType;
  color?: string;
  icon?: string;
}

// Transaction Category update interface
export interface TransactionCategoryUpdate {
  name?: string;
  type?: TransactionCategoryType;
  color?: string;
  icon?: string;
}

// Transaction Category with subcategories
export interface TransactionCategoryWithSubcategories extends TransactionCategory {
  subcategories: TransactionSubcategory[];
}

// Transaction Subcategory interface
export interface TransactionSubcategory {
  id: number;
  tenant_id: number;
  user_id?: number;
  category_id: number;
  name: string;
  color?: string;
  icon?: string;
  is_system: boolean;
  created_at: string; // ISO string
}

// Transaction Subcategory creation interface
export interface TransactionSubcategoryCreate {
  category_id: number;
  name: string;
  color?: string;
  icon?: string;
}

// Transaction Subcategory update interface
export interface TransactionSubcategoryUpdate {
  name?: string;
  color?: string;
  icon?: string;
}

// Financial Account interface
export interface FinancialAccount {
  id: number;
  tenant_id: number;
  user_id: number;
  account_name: string;
  account_type?: string;
  account_number?: string;
  bank_name?: string;
  current_balance: number; // Decimal from backend as number in frontend
  currency: string;
  is_active: boolean;
  created_at: string; // ISO string
  updated_at: string; // ISO string
}

// Financial Account creation interface
export interface FinancialAccountCreate {
  account_name: string;
  account_type?: string;
  account_number?: string;
  bank_name?: string;
  current_balance?: number;
  currency?: string;
  is_active?: boolean;
}

// Financial Account update interface
export interface FinancialAccountUpdate {
  account_name?: string;
  account_type?: string;
  account_number?: string;
  bank_name?: string;
  current_balance?: number;
  currency?: string;
  is_active?: boolean;
}

// Financial Account with summary
export interface FinancialAccountWithSummary extends FinancialAccount {
  transaction_count: number;
  total_credits: number;
  total_debits: number;
  last_transaction_date?: string; // ISO string
}

// Recurring Transaction interface
export interface RecurringTransaction {
  id: number;
  tenant_id: number;
  user_id: number;
  template_name: string;
  type: TransactionType;
  amount: number; // Decimal from backend as number in frontend
  currency: string;
  description?: string;
  category_id?: number;
  subcategory_id?: number;
  account_id?: number;
  frequency: RecurringFrequency;
  interval_count: number;
  start_date: string; // ISO date string
  end_date?: string; // ISO date string
  reminder_days: number;
  is_active: boolean;
  next_due_date?: string; // ISO date string
  extra_data?: Record<string, unknown>;
  created_at: string; // ISO string
  updated_at: string; // ISO string
}

// Recurring Transaction creation interface
export interface RecurringTransactionCreate {
  template_name: string;
  type: TransactionType;
  amount: number;
  currency?: string;
  description?: string;
  category_id?: number;
  subcategory_id?: number;
  account_id?: number;
  frequency: RecurringFrequency;
  interval_count?: number;
  start_date: string; // ISO date string
  end_date?: string; // ISO date string
  reminder_days?: number;
  is_active?: boolean;
  extra_data?: Record<string, unknown>;
}

// Recurring Transaction update interface
export interface RecurringTransactionUpdate {
  template_name?: string;
  type?: TransactionType;
  amount?: number;
  currency?: string;
  description?: string;
  category_id?: number;
  subcategory_id?: number;
  account_id?: number;
  frequency?: RecurringFrequency;
  interval_count?: number;
  start_date?: string; // ISO date string
  end_date?: string; // ISO date string
  reminder_days?: number;
  is_active?: boolean;
  extra_data?: Record<string, unknown>;
}

// Recurring Transaction with next occurrence info
export interface RecurringTransactionWithNext extends RecurringTransaction {
  next_occurrence_date?: string; // ISO date string
  days_until_next?: number;
}

// Filtering interfaces
export interface TransactionFilters {
  type?: TransactionType;
  category_id?: number;
  subcategory_id?: number;
  account_id?: number;
  date_from?: string; // ISO date string
  date_to?: string; // ISO date string
  amount_min?: number;
  amount_max?: number;
  currency?: string;
  description?: string;
  is_recurring?: boolean;
}

// Search interface
export interface TransactionSearchQuery {
  query: string;
  filters?: TransactionFilters;
  sort_by?: 'date' | 'amount' | 'description' | 'created_at';
  sort_order?: 'asc' | 'desc';
  limit?: number;
  offset?: number;
}

// Bulk operations
export interface TransactionBulkUpdate {
  transaction_ids: number[];
  category_id?: number;
  subcategory_id?: number;
  account_id?: number;
}

// Constants for UI
export const TRANSACTION_TYPE_OPTIONS: { value: TransactionType; label: string; color: string }[] = [
  { value: 'credit', label: 'Credit (Income)', color: 'green' },
  { value: 'debit', label: 'Debit (Expense)', color: 'red' },
];

export const TRANSACTION_CATEGORY_TYPE_OPTIONS: { value: TransactionCategoryType; label: string; color: string }[] = [
  { value: 'income', label: 'Income', color: 'green' },
  { value: 'expense', label: 'Expense', color: 'red' },
  { value: 'transfer', label: 'Transfer', color: 'blue' },
];

export const RECURRING_FREQUENCY_OPTIONS: { value: RecurringFrequency; label: string }[] = [
  { value: 'daily', label: 'Daily' },
  { value: 'weekly', label: 'Weekly' },
  { value: 'monthly', label: 'Monthly' },
  { value: 'quarterly', label: 'Quarterly' },
  { value: 'yearly', label: 'Yearly' },
];

// Additional interfaces for state management
export interface TransactionSortBy {
  field: 'date' | 'amount' | 'description' | 'created_at';
  direction: 'asc' | 'desc';
}

export interface TransactionFormData {
  type: TransactionType;
  amount: string; // String for form input
  currency: string;
  description: string;
  transaction_date: string;
  category_id: string; // String for form select
  subcategory_id: string; // String for form select
  account_id: string; // String for form select
  reference_number: string;
  is_recurring: boolean;
  recurring_template_id: string; // String for form select
}

export interface ValidationErrors {
  type?: string;
  amount?: string;
  currency?: string;
  description?: string;
  transaction_date?: string;
  category_id?: string;
  subcategory_id?: string;
  account_id?: string;
  reference_number?: string;
  general?: string;
}

export type CreateTransactionInput = TransactionCreate;
export type CreateCategoryInput = TransactionCategoryCreate;
export type CreateAccountInput = FinancialAccountCreate;
export type CreateRecurringTransactionInput = RecurringTransactionCreate;

// View mode for transaction display
export type TransactionViewMode = 'list' | 'table' | 'calendar' | 'chart';

// Transaction summary interfaces
export interface TransactionSummary {
  total_income: number;
  total_expenses: number;
  net_amount: number;
  transaction_count: number;
  currency: string;
  period_start: string; // ISO date string
  period_end: string; // ISO date string
}

export interface CategorySummary {
  category_id: number;
  category_name: string;
  total_amount: number;
  transaction_count: number;
  percentage: number;
}

export interface AccountSummary {
  account_id: number;
  account_name: string;
  current_balance: number;
  total_credits: number;
  total_debits: number;
  transaction_count: number;
  currency: string;
}